### ISSUE TITLE
BrightID SDK - Monorepo Architecture

### ISSUE THREAD

**Issue by alimahdiyar** [BrightID/BrightID#1199]
Hi

There are some react codes in BrightID mobile application that are needed for all SuperUserApps, including the import flow and BrightID operations. For Aura as a SuperUserApp, I copied all the codes I needed from BrightID's source code like this:

https://github.com/Meta-Node/aura-frontend/tree/master/src/BrightID

But to keep the code clean and avoid the duplicates, it is better if we extract the codes that provide the core functionality of BrightID as an SDK which can be used by both the BrightID Client and the SuperUserApps.

So if we do this, we're going to have a "superuser-sdk" package and a "brightid-client" package which uses "superuser-sdk" to work. People who are going to work on "brightid-client" should frequently update the "superuser-sdk" and see the result on the brightid client, so it makes sense to bring these two packages under one GitHub repository under a monorepo architecture. This can be done in Node.js package managers by using "workspace"s.

Workspaces are supported by pnpm and new versions of yarn. I tried updating yarn or migrating to pnpm for BrightID mobile client, but I couldn't fix the build. Here is an example of a monorepo project using pnpm:
https://github.com/nucypher/taco-web
As you can see in the repo, their react example in the repo is using `@nucypher/taco` package:
https://github.com/nucypher/taco-web/blob/ca5c19d43da968f869b2c80df8fdace3e36ee76e/examples/taco/react/src/App.tsx#L2
Which is provided by another package in the repository like this:
https://github.com/nucypher/taco-web/blob/ca5c19d43da968f869b2c80df8fdace3e36ee76e/examples/taco/react/package.json#L27

I've heard in the meeting that there are ongoing efforts to upgrade BrigthID mobile packages, like the PR opened by @youngkidwarrior. So I guess it's worth it if we also migrate the source code to a this new structure while upgrading the packages.

I'm relatively new to BrightID mobile app development, and can't test BrightID client's features with the same quality as the old devs. So I wrote this report so that some devs might be able to implement this new structure

**Comment by adamstallard** [comment]
I don't know if it makes more sense to have two different SDKs--one for superapp functionality and one for regular functionality.  It might be less confusing to combine those functions into a single BrightID SDK and apps can choose whether or not they need the superapp functions.

**Comment by adamstallard** [comment]
Another point about monorepos is the last time I used one with clrfund, keeping the different changes and packages in sync within the monorepo was harder and caused more problems than if we would have had separate repos.

Supposedly, the tooling has gotten better--- pnpm now has "workspaces." [The documentation on it](https://pnpm.io/workspaces) recommends using "bit" for dependency management and "changeset" or "Rush" for versioning.

If learning and using these new tools is going to be easier than managing two repos, then let's do it.

**Comment by adamstallard** [comment]
There's also another issue about outside contributions.  If it's going to make it harder for outside contributors to update the BrightID javascript SDK because it's inside a monorepo instead of in its own, then we shouldn't do it.

**Comment by TripleSpeeder** [comment]
Some thoughts from my side, not having worked much with monorepos so far:


TL;DR: I think monorepos are fine. But I'm not sure if it makes sense for our particular usecase.


If we launch a dedicated SDK (which I'm strongly in favor of) for talking to our Node API we will need to be very careful about backwards compatibility and correct versioning. 

If one of the monorepos benefits is the easy way to update/release dependencies and client at the same time/in sync, this seems to be not a big advantage for our case - We can not release breaking changes anyway. 

> There's also another issue about outside contributions. If it's going to make it harder for outside contributors to update the BrightID javascript SDK because it's inside a monorepo instead of in its own repo, then we shouldn't do it.

I think this is a very good point. The BrightID client codebase is huge, confusing and complex. I would much prefer to have a clean start for an SDK without any additional code in the repo. I'm afraid that new developers interested in BrightID will be intimidated by the whole client code. 

Overall I kind of like the scenario where the SDK and mobile client are rather loosely coupled. I can see a future where the mobile client will be less important and may be replaced by third-party web apps or other more light-weight implementations. 

> People who are going to work on "brightid-client" should frequently update the "superuser-sdk" and see the result on the brightid client,

Although not a very popular feature, i think for this case using `npm link` functionality should also work okay. I used this in the past for other projects and had no major issues.

**Comment by alimahdiyar** [comment]
@adamstallard @TripleSpeeder The flow for the new developers does not change. For them, different packages are just different folders and they don't need to know anything about workspaces to be able to work with that

**Comment by alimahdiyar** [comment]
@TripleSpeeder Thanks for suggesting `npm link`. That could work in our scenario too. But still, people should still remember to pull sdk git repository every time. If we put these both in a repo but without using workspaces, that might work, too

**Comment by adamstallard** [comment]
Some of the refactors leading toward a monorepo architecture are already in PRs #1202 and #1203
